src/dot_utils.rs

```diff
@@ -0,0 +1,85 @@
+// Licensed under the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License. You may obtain
+// a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+// WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+// License for the specific language governing permissions and limitations
+// under the License.
+
+use std::collections::BTreeMap;
+use std::io::prelude::*;
+
+use petgraph::visit::{
+    Data, EdgeRef, GraphBase, GraphProp, IntoEdgeReferences,
+    IntoNodeReferences, NodeIndexable, NodeRef,
+};
+use pyo3::prelude::*;
+
+static TYPE: [&str; 2] = ["graph", "digraph"];
+static EDGE: [&str; 2] = ["--", "->"];
+
+pub fn build_dot<G, N, E, T>(
+    graph: G,
+    file: &mut T,
+    graph_attrs: Option<BTreeMap<String, String>>,
+    mut node_attrs: N,
+    mut edge_attrs: E,
+) -> PyResult<()>
+where
+    T: Write,
+    G: GraphBase
+        + IntoEdgeReferences
+        + IntoNodeReferences
+        + NodeIndexable
+        + GraphProp,
+    G: Data<NodeWeight = PyObject, EdgeWeight = PyObject>,
+    N: FnMut(&PyObject) -> PyResult<BTreeMap<String, String>>,
+    E: FnMut(&PyObject) -> PyResult<BTreeMap<String, String>>,
```

Comment:
I don't follow what's going on with G N E T here

Reply:
These are [bounds](https://doc.rust-lang.org/stable/rust-by-example/generics/bounds.html) for generic types on this function. Basically it's saying that the parameter `graph` is of generic type `G` where any object passed in via graph implements a bunch of traits. A [trait](https://doc.rust-lang.org/book/ch10-02-traits.html) defines a set of common methods and attributes that can be implemented for a type. The specific list of traits for `G` are from petgraph and basically are just shared functions for graph traversal and access (I got most of them from compiler errors and the similar function in petgraph that generates dot). The definitions of them are here: https://docs.rs/petgraph/0.5.1/petgraph/visit/index.html#traits but they basically just enable us to pass in either a `PyGraph` or `PyDiGraph` and have the function work as expected for both (the trait definitions for either type are at the top of both `digraph.rs` and `graph.rs`).

`T`, `N`, and `E` are easier to follow because they're much simpler. Parameter `file` is a mutable borrow of generic type `T`, where `T` is anything that implements the [Write](https://doc.rust-lang.org/std/io/trait.Write.html) trait. This can be a file object, a buffer, etc. Parameter `node_attrs` is a generic type `N` whiere `N` is any function that takes in a `&PyObject` and returns a `PyResult<BTreeMap<String, String>` and parameter `edge_attrs` is a generic type `E` where it's also a function like that (`E` and `N` are different though because they are not necessarily going to be the same function).